### PR TITLE
Car demo with events

### PR DESCRIPTION
**Pull request by MaGaMeGa**

### CODE REVIEW

File: demo-car-with-events/README.md [hunk 1]
```diff
@@ -0,0 +1,84 @@
+# Car SoS Demo With Events (Java Spring-Boot)
```
Comment (borditamas) [inline demo-car-with-events/README.md]: Highlight and link it in the main ReadMe

File: demo-car-with-events/demo-car-consumer-with-subscribing/src/main/java/ai/aitia/demo/car_consumer_with_subscribing/CarConsumerWithSubscriptionApplicationInitListener.java [hunk 2]
```diff
@@ -0,0 +1,242 @@
...
+			final PublicKey authorizationPublicKey = arrowheadService.queryAuthorizationPublicKey();
+			if (authorizationPublicKey == null) {
+				throw new ArrowheadException("Authorization public key is null");
+			}
+			
+			KeyStore keystore;
+			try {
+				keystore = KeyStore.getInstance(sslProperties.getKeyStoreType());
+				keystore.load(sslProperties.getKeyStore().getInputStream(), sslProperties.getKeyStorePassword().toCharArray());
+			} catch (KeyStoreException | NoSuchAlgorithmException | CertificateException | IOException ex) {
```
Comment (rbocsi) [inline demo-car-with-events/demo-car-consumer-with-subscribing/src/main/java/ai/aitia/demo/car_consumer_with_subscribing/CarConsumerWithSubscriptionApplicationInitListener.java]: missing final

File: demo-car/demo-car-provider/src/main/resources/application.properties [hunk 3]
```diff
@@ -22,7 +22,7 @@ sr_port=8443
 # Set this to false to disable https mode
 server.ssl.enabled=true
 # Set this to false to deactivate the token security filter
-token.security.filter.enabled=true
+token.security.filter.enabled=false
```
Comment (borditamas) [inline demo-car/demo-car-provider/src/main/resources/application.properties]: set back to true

File: demo-car-with-events/demo-car-consumer-with-subscribing/src/main/java/ai/aitia/demo/car_consumer_with_subscribing/CarConsumerWithSubscriptionController.java [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+
+	//-------------------------------------------------------------------------------------------------
+	@GetMapping(path = CommonConstants.ECHO_URI)
+	public String echoService() {
+		return "Got it!";
+	}
+	
+	//-------------------------------------------------------------------------------------------------
+	@PostMapping(path = SubscriberConstants.REQUEST_RECEIVED_NOTIFICATION_URI) 
+	public void receivePublisherReceivedRequestEvent(@RequestBody final EventDTO event ) {
```
Comment (rbocsi) [inline demo-car-with-events/demo-car-consumer-with-subscribing/src/main/java/ai/aitia/demo/car_consumer_with_subscribing/CarConsumerWithSubscriptionController.java]: rename receiveEventRequestReceived

File: demo-car-with-events/demo-car-consumer-with-subscribing/src/main/java/ai/aitia/demo/car_consumer_with_subscribing/CarConsumerWithSubscriptionController.java [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+
+		}
+		
+		
+		//TODO implement your event handling logic here 
+	}
+	
+	//-------------------------------------------------------------------------------------------------
+	@PostMapping(path = SubscriberConstants.PUBLISHER_DESTORYED_NOTIFICATION_URI) 
+	public void receivePublsisherDestroyedEvent(@RequestBody final EventDTO event ) {
```
Comment (rbocsi) [inline demo-car-with-events/demo-car-consumer-with-subscribing/src/main/java/ai/aitia/demo/car_consumer_with_subscribing/CarConsumerWithSubscriptionController.java]: rename receiveEventDestroyed

File: demo-car/demo-car-provider/src/main/resources/application.properties [hunk 6]
```diff
@@ -7,7 +7,7 @@ client_system_name=car_provider_demo
 
 # Client web-server parameters
 server.address=127.0.0.1
-server.port=8888
+server.port=9888
```
Comment (borditamas) [inline demo-car/demo-car-provider/src/main/resources/application.properties]: set back to 8888

File: demo-car-with-events/demo-car-consumer-with-subscribing/src/main/java/ai/aitia/demo/car_consumer_with_subscribing/CarConsumerWithSubscriptionTask.java [hunk 7]
```diff
@@ -0,0 +1,379 @@
...
+	// members
+	
+	private boolean interrupted = false;
+	
+	private final Logger logger = LogManager.getLogger(CarConsumerWithSubscriptionTask.class);
+	
+	@Resource( name = SubscriberConstants.NOTIFICATION_QUEUE )
+	private ConcurrentLinkedQueue<EventDTO> notificatonQueue;
+	
+	final Set<String> receivedEvenTypeList = new HashSet<>();
```
Comment (rbocsi) [inline demo-car-with-events/demo-car-consumer-with-subscribing/src/main/java/ai/aitia/demo/car_consumer_with_subscribing/CarConsumerWithSubscriptionTask.java]: missing private

File: demo-car-with-events/demo-car-consumer-with-subscribing/src/main/java/ai/aitia/demo/car_consumer_with_subscribing/CarConsumerWithSubscriptionTask.java [hunk 8]
```diff
@@ -0,0 +1,379 @@
...
+
+		OrchestrationResultDTO carCreationService = null;
+		OrchestrationResultDTO carRequestingService = null;
+		
+		int counter = 0;
+		while ( !interrupted && ( counter < max_retry ) ) {
+			
+			try {
+				
+				if ( !( notificatonQueue.peek() == null) ) {
```
Comment (rbocsi) [inline demo-car-with-events/demo-car-consumer-with-subscribing/src/main/java/ai/aitia/demo/car_consumer_with_subscribing/CarConsumerWithSubscriptionTask.java]: use !=

File: demo-car-with-events/demo-car-provider-with-publishing/src/main/java/ai/aitia/demo/car_provider_with_publishing/CarProviderWithPublishingApplicationInitListener.java [hunk 9]
```diff
@@ -0,0 +1,196 @@
...
+
+		}
+		
+	}
+	
+	//-------------------------------------------------------------------------------------------------
+	@Override
+	public void customDestroy() {
+		//Unregister service
+		publishDestoyedEvent();
```
Comment (rbocsi) [inline demo-car-with-events/demo-car-provider-with-publishing/src/main/java/ai/aitia/demo/car_provider_with_publishing/CarProviderWithPublishingApplicationInitListener.java]: typo destRoyedEvent

File: demo-car-with-events/demo-car-provider-with-publishing/src/main/java/eu/arrowhead/client/skeleton/publisher/constants/PublisherConstants.java [hunk 10]
```diff
@@ -0,0 +1,19 @@
+package eu.arrowhead.client.skeleton.publisher.constants;
+
+public class PublisherConstants {
+	//=================================================================================================
+	// members
+
+	public static final String START_INIT_EVENT_PAYLOAD= "InitStarted";
+	public static final String START_RUN_EVENT_PAYLOAD= "RunStarted";
+	public static final String PUBLISHR_DESTROYED_EVENT_PAYLOAD= "RunStarted";
```
Comment (rbocsi) [inline demo-car-with-events/demo-car-provider-with-publishing/src/main/java/eu/arrowhead/client/skeleton/publisher/constants/PublisherConstants.java]: ???

File: demo-car-with-events/demo-car-provider-with-publishing/src/main/resources/META-INF/additional-spring-configuration-metadata.json [hunk 11]
```diff
@@ -0,0 +1,22 @@
...
+    "description": "A description for 'sr_address'"
+  },
+  {
+    "name": "sr_port",
+    "type": "java.lang.String",
+    "description": "A description for 'sr_port'"
+  },
+  {
+    "name": "token.security.filter.enabled",
+    "type": "java.lang.String",
```
Comment (rbocsi) [inline demo-car-with-events/demo-car-provider-with-publishing/src/main/resources/META-INF/additional-spring-configuration-metadata.json]: Boolean

File: demo-car/demo-car-consumer/src/main/java/ai/aitia/demo/car_consumer/CarConsumerMain.java [hunk 12]
```diff
@@ -67,8 +67,9 @@ public void createCarServiceOrchestrationAndConsumption() {
     	
 		final Builder orchestrationFormBuilder = arrowheadService.getOrchestrationFormBuilder();
 		final OrchestrationFormRequestDTO orchestrationFormRequest = orchestrationFormBuilder.requestedService(serviceQueryForm)
-																					   .flag(Flag.MATCHMAKING, true)
+																					   .flag(Flag.MATCHMAKING, false)
 																					   .flag(Flag.OVERRIDE_STORE, true)
+																					   .flag(Flag.PING_PROVIDERS, true)
```
Comment (rbocsi) [inline demo-car/demo-car-consumer/src/main/java/ai/aitia/demo/car_consumer/CarConsumerMain.java]: revert this file

File: demo-car/demo-car-provider/.springBeans [hunk 13]
```diff
@@ -7,7 +7,7 @@
 	</configSuffixes>
 	<enableImports><![CDATA[false]]></enableImports>
 	<configs>
-		<config>java:ai.aitia.demo.car_provider.CarProviderMain</config>
```
[the remark this answers is not in the document]
Comment (MaGaMeGa) [reply demo-car/demo-car-provider/.springBeans]: fixed

File: demo-car/demo-car-consumer/src/main/java/ai/aitia/demo/car_consumer/CarConsumerMain.java (hunk 12 above)
Comment (MaGaMeGa) [reply demo-car/demo-car-consumer/src/main/java/ai/aitia/demo/car_consumer/CarConsumerMain.java]: fixed

File: demo-car-with-events/demo-car-provider-with-publishing/src/main/resources/META-INF/additional-spring-configuration-metadata.json (hunk 11 above)
Comment (MaGaMeGa) [reply demo-car-with-events/demo-car-provider-with-publishing/src/main/resources/META-INF/additional-spring-configuration-metadata.json]: fixed

File: demo-car-with-events/demo-car-provider-with-publishing/src/main/java/eu/arrowhead/client/skeleton/publisher/constants/PublisherConstants.java (hunk 10 above)
Comment (MaGaMeGa) [reply demo-car-with-events/demo-car-provider-with-publishing/src/main/java/eu/arrowhead/client/skeleton/publisher/constants/PublisherConstants.java]: fixed

File: demo-car-with-events/demo-car-consumer-with-subscribing/src/main/java/ai/aitia/demo/car_consumer_with_subscribing/CarConsumerWithSubscriptionTask.java [hunk 14]
```diff
@@ -0,0 +1,379 @@
...
+			
+			arrowheadService.subscribeToEventHandler( subscription );
+		
+		} catch ( final InvalidParameterException ex) {
+			
+			if( ex.getMessage().contains( "Subscription violates uniqueConstraint rules" )) {
+				
+				logger.debug("Subscription is already in DB");
+			}
+
```
[the remark this answers is not in the document]
Comment (MaGaMeGa) [reply demo-car-with-events/demo-car-consumer-with-subscribing/src/main/java/ai/aitia/demo/car_consumer_with_subscribing/CarConsumerWithSubscriptionTask.java]: fixed

File: demo-car-with-events/demo-car-consumer-with-subscribing/src/main/java/ai/aitia/demo/car_consumer_with_subscribing/CarConsumerWithSubscriptionTask.java (hunk 8 above)
Comment (MaGaMeGa) [reply demo-car-with-events/demo-car-consumer-with-subscribing/src/main/java/ai/aitia/demo/car_consumer_with_subscribing/CarConsumerWithSubscriptionTask.java]: fixed

File: demo-car-with-events/demo-car-consumer-with-subscribing/src/main/java/ai/aitia/demo/car_consumer_with_subscribing/CarConsumerWithSubscriptionTask.java (hunk 7 above)
Comment (MaGaMeGa) [reply demo-car-with-events/demo-car-consumer-with-subscribing/src/main/java/ai/aitia/demo/car_consumer_with_subscribing/CarConsumerWithSubscriptionTask.java]: removed

File: demo-car/demo-car-provider/src/main/resources/application.properties (hunk 6 above)
Comment (MaGaMeGa) [reply demo-car/demo-car-provider/src/main/resources/application.properties]: fixed

File: demo-car-with-events/demo-car-consumer-with-subscribing/src/main/java/ai/aitia/demo/car_consumer_with_subscribing/CarConsumerWithSubscriptionController.java (hunk 5 above)
Comment (MaGaMeGa) [reply demo-car-with-events/demo-car-consumer-with-subscribing/src/main/java/ai/aitia/demo/car_consumer_with_subscribing/CarConsumerWithSubscriptionController.java]: fixed

File: demo-car-with-events/demo-car-consumer-with-subscribing/src/main/java/ai/aitia/demo/car_consumer_with_subscribing/CarConsumerWithSubscriptionController.java (hunk 4 above)
Comment (MaGaMeGa) [reply demo-car-with-events/demo-car-consumer-with-subscribing/src/main/java/ai/aitia/demo/car_consumer_with_subscribing/CarConsumerWithSubscriptionController.java]: fixed

File: demo-car-with-events/demo-car-consumer-with-subscribing/src/main/java/ai/aitia/demo/car_consumer_with_subscribing/CarConsumerWithSubscriptionApplicationInitListener.java (hunk 2 above)
Comment (MaGaMeGa) [reply demo-car-with-events/demo-car-consumer-with-subscribing/src/main/java/ai/aitia/demo/car_consumer_with_subscribing/CarConsumerWithSubscriptionApplicationInitListener.java]: fixed

File: demo-car/demo-car-provider/src/main/resources/application.properties (hunk 3 above)
Comment (MaGaMeGa) [reply demo-car/demo-car-provider/src/main/resources/application.properties]: fixed

**Pull request merged**